### ISSUE TITLE
Cannot specify file or directory with extra CAs

### ISSUE THREAD

**Issue by keestux** [bitwarden/clients#5075]
### Steps To Reproduce

# Prereqs

- bitwarden self hosted server with a SSL certificate issued by a FreeIPA self-signed CA

# Steps to reproduce:
1. open the desktop app
2. change config with local URL of bitwarden server
3. enter email and password

### Expected Result

The client should normally connect to the server

### Actual Result

Pop up with "An error has occurred. Failed to fetch"

If I start `bitwarden` from the command line I see:
```
[274681:0324/155037.087913:ERROR:nss_util.cc(349)] After loading Root Certs, loaded==false: NSS error code: -8018
[274681:0324/155037.088659:ERROR:cert_verify_proc_builtin.cc(670)] CertVerifyProcBuiltin for vaultwarden.example.com failed:
----- Certificate i=0 (CN=vaultwarden.example.com,O=EXAMPLE.COM) -----
ERROR: No matching issuer found
```

### Additional Context

The cli tool `bw` has the same problem. However I found that it looks at `NODE_EXTRA_CA_CERTS` envvar.

Without the envvar
```
$ bw login kees@example.com
? Master password: [hidden]
request to https://vaultwarden.example.com/identity/accounts/prelogin failed, reason: unable to verify the first certificate
```

With the envvar
```
$ NODE_EXTRA_CA_CERTS=/etc/ipa/ca.crt bw login kees@example.com
? Master password: [hidden]
You are logged in!

To unlock your vault, set your session key to the `BW_SESSION` environment variable. ex:
$ export BW_SESSION="xxxxxxxxxxxx=="
> $env:BW_SESSION="xxxxxxxxxxxx=="

You can also pass the session key to any command with the `--session` option. ex:
$ bw list items --session xxxxxxxxxxxx==
```

Unfortunately that envvar is not used by the desktop app.

### Operating System

Linux

### Operating System Version

Ubuntu 20.04

### Installation method

Other

### Build Version

2023.3.0

### Issue Tracking Info

- [X] I understand that work is tracked outside of Github. A PR will be linked to this issue should one be opened to address it, but Bitwarden doesn't use fields like "assigned", "milestone", or "project" to track progress.

**Comment by SergeantConfused** [comment]
Hi @keestux,

Thank you for your report. It sounds like you need to trust that self-signed certificate as outlined here (https://bitwarden.com/help/certificates/#trust-a-self-signed-certificate); If you run into any trouble with that, please feel free to create a support ticket through our Help Centre (https://bitwarden.com/help/) and we'll assist you.

By the way, it seems like you're using a vaultwarden (bitwarden_rs) server (https://github.com/dani-garcia/vaultwarden), not a Bitwarden server.

We use GitHub issues as a place to track bugs and other development related issues; I am going to close this report now.


Thank you again,

**Comment by keestux** [comment]
Hey @SergeantConfused it is a pity you already closed this issue. I really believe there is a problem with the desktop client.

I already looked at that help information about self-signed certificates. The CA of the FreeIPA server is already in the system `ca-certificates`.

Above I showed you that for the command line version (`bw`) there is a work around, i.e. set the envvar `NODE_EXTRA_CA_CERTS`. However the desktop client does not use it. It seems to be using its built-in list of trusted CAs (not sure where it gets that from).

Have a look at https://github.com/electron/electron/issues/10257 where it is being discussed that electron does not look at `NODE_EXTRA_CA_CERTS` out of the box. You have to add some extra code for it.

PS. Yes, I have a vaultwarden server, but this issue is about the bitwarden desktop client.

**Comment by joshtrichards** [comment]
@keestux I believe the desktop client uses Electron's support for Chromium networking as an alternative to the native Node.js network lib[1]. And I think is why is the desktop app requires a different process beyond `ca-certificates`. It's easy to miss *after* the `ca-certificates` stuff in the docs, but it's there:

> For our Linux desktop app, accessing the web vault using Chromium-based browsers, and the Directory Connector desktop app, you also need to complete this Linux cert management procedure: https://chromium.googlesource.com/chromium/src/+/refs/heads/master/docs/linux/cert_management.md

On Ubuntu this entails installing `libnss3-tools` then using `certutil` to add the cert (more details at the above chromium url). Unfortunately I don't have an easy way to test this right at the moment or I'd confirm it works as documented. 

As for the CLI, yep that all sounds right and matches what's documented here: https://bitwarden.com/help/certificates/#trust-a-self-signed-certificate (where the above excerpt can also be found for the desktop specific process).

There may be some room for streamlining the differences between the apps/platforms from the perspective of day to day users, but there is a documented approach that is supposed to still work (Disclaimer: I do _not_ work for BW)

[1] https://www.electronjs.org/docs/latest/api/net

**Comment by keestux** [comment]
@joshtrichards thanks for the pointer. You've put me on the right track.

Indeed, I completely ignored that description of chromium-based browsers, because I'm not really using it.

I started chromium and added my CA, but that didn't help. That `certutil` command in the doc also added to the confusion, because you shouldn't be using the `sql:` prefix. The command to list chromium's certificates must be:
```
certutil -d $HOME/.pki/nssdb -L
```
My added CA cert wasn't there. What? Oh wait, I have chromium snap, which stores the nssdb in `~/snap/chromium/current/.pki/nssdb` :-(
There was a `~/.pki/nssdb` from the time chromium was a non-snap application and before I added my CA cert to it.

My current work-around is to copy the pki from chromium's snap
```
cp -a ~/snap/chromium/current/.pki ~/
```

Now it works. Gee, that was quite an adventure.
